File: 1872-stone-game-viii/solution.cpp
#include <vector>
#include <numeric>
#include <algorithm>

class Solution {
public:
    int stoneGameVIII(std::vector<int>& stones) {
        int n = stones.size();
        
        // Compute the total sum of the array (which is pref[n-1])
        long long pref_sum = 0;
        for (int stone : stones) {
            pref_sum += stone;
        }

        // dp stores the maximum score difference starting from the current index.
        // Base case: dp[n-1] = pref[n-1]
        long long dp = pref_sum;

        // Traverse backwards from index n-1 down to 2
        // Update pref_sum to be pref[i-1] for each step
        for (int i = n - 1; i >= 2; --i) {
            pref_sum -= stones[i];
            dp = std::max(dp, pref_sum - dp);
        }

        return dp;
    }
};

File: 1872-stone-game-viii/solution.py
class Solution:
    def stoneGameVIII(self, stones: list[int]) -> int:
        n = len(stones)
        
        # Calculate total sum of all stones (which represents pref[n-1])
        pref_sum = sum(stones)
        
        # Base case: dp[n-1] = pref[n-1]
        dp = pref_sum
        
        # Iterate backwards from index n-1 down to 2
        # In each step, adjust pref_sum to represent pref[i-1]
        for i in range(n - 1, 1, -1):
            pref_sum -= stones[i]
            dp = max(dp, pref_sum - dp)
            
        return dp